src/background.js

```diff
@@ -4,62 +4,303 @@
+    setTimeout(() => {
+      currentTip = tip;
+      browser.urlbar.search("", { focus: tip == TIPS.ONBOARD });
+    }, SHOW_TIP_DELAY_MS);
+  }
+}
+
+/**
+ * browser.urlbar.onBehaviorRequested listener.
+ */
+async function onBehaviorRequested(query) {
+  return currentTip ? "restricting" : "inactive";
+}
+
+/**
+ * browser.urlbar.onResultsRequested listener.
+ */
+async function onResultsRequested(query) {
+  let tip = currentTip;
+  currentTip = TIPS.NONE;
+
+  let engines = await browser.search.get();
+  let defaultEngine = engines.find(engine => engine.isDefault);
+
+  let result = {
+    type: "tip",
+    source: "local",
+    heuristic: true,
+    payload: {
+      icon: defaultEngine.favIconUrl,
+      buttonText: "Okay, Got It",
+    },
+  };
+
+  switch (tip) {
+    case TIPS.ONBOARD:
+      result.payload.text =
+        `Type less, find more: Search ${defaultEngine.name} ` +
+        `right from your address bar.`;
+      break;
+    case TIPS.REDIRECT:
+      result.payload.text =
+        `Start your search here to see suggestions from ` +
+        `${defaultEngine.name} and your browsing history.`;
+      break;
+  }
+
+  return [result];
+}
+
+/**
+ * browser.urlbar.onResultPicked listener.  Called when a tip button is picked.
+ */
+async function onResultPicked(payload) {
+  browser.urlbar.focus();
+
+  // The user clicked the "Okay, Got It" button.  We shouldn't show a tip again
+  // in any session.  Set the shown count to the max.
+  storage[STORAGE_KEY_SHOWN_COUNT] = MAX_SHOWN_COUNT;
+  await browser.storage.local.set(storage);
+}
+
+/**
+ * browser.webNavigation.onBeforeNavigate listener.  Called when a new
+ * navigation starts.  We use this to close the urlbar view, which is necessary
+ * when the input isn't focused.
+ */
+async function onBeforeNavigate(details) {
+  // frameId == 0 for top-level loads.
+  if (details.frameId == 0) {
+    let tab = await browser.tabs.get(details.tabId);
+    if (tab.active) {
+      browser.urlbar.closeView();
+    }
+  }
+}
+
+/**
+ * browser.windows.onFocusChanged listener.  We use this to close the urlbar
+ * view, which is necessary when the input isn't focused.
+ */
+function onWindowFocusChanged() {
+  browser.urlbar.closeView();
+}
+
+/**
+ * Resets all the state we set on enrollment in the study.
+ */
+async function unenroll() {
+  await browser.tabs.onActivated.removeListener(onTabActivated);
+  await browser.webNavigation.onCompleted.removeListener(onWebNavigation);
+  await browser.urlbar.onBehaviorRequested.removeListener(onBehaviorRequested);
+  await browser.urlbar.onResultsRequested.removeListener(onResultsRequested);
+  await browser.urlbar.onResultPicked.removeListener(onResultPicked);
+  await browser.webNavigation.onBeforeNavigate.removeListener(onBeforeNavigate);
+  await browser.windows.onFocusChanged.removeListener(onWindowFocusChanged);
+  await browser.experiments.urlbar.engagementTelemetry.clear({});
```

Comment:
IT may be safer to do this as first thing, and remove listeners later.

Reply:
Seems reasonable, I'll update.